src/lib/InstantSearch.ts

```diff
@@ -80,6 +81,17 @@ export type InstantSearchOptions<TRouteState = UiState> = {
    */
   searchFunction?: (helper: AlgoliaSearchHelper) => void;
 
+  /**
+   * Function called when the state changes.
+   *
+   * Using this function turns the experience controlled. This means that you
+   * become in charge of updating the UI state with the `setUiState` function.
```

Comment:
What happens if you call the function async. Is that something you tried?

Reply:
https://codesandbox.io/s/instantsearchjs-controlled-mode-so5qw, looks like it makes a global debounce possible, but seems a bit glitchy to me 🤔

Reply:
It behaves as I would expect to: [see sandbox](https://codesandbox.io/s/instantsearchjs-controlled-mode-3c26u). What seems glitchy to you?

Reply:
all updates happen still at the same speed as you initially did, but they are delayed. It's clearer to see with a higher value for the timeout.

I wonder if the older values maybe need to be cancelled, I'll try that

Reply:
https://codesandbox.io/s/instantsearchjs-controlled-mode-so5qw, yes this way it's way smoother. 

```js
let timeoutHandle;

const search = instantsearch({
  indexName: 'instant_search',
  searchClient,
  // Try controlled mode here!
  onStateChange({ uiState, setUiState }) {
    console.group('onStateChange');
    console.log({ uiState });
    console.groupEnd();

    if (timeoutHandle) {
      clearTimeout(timeoutHandle);
    }

    timeoutHandle = setTimeout(() => setUiState(uiState), 500);
  },
});
```

If we document this in the future, we need to make sure to cancel the stale updates. I don't think InstantSearch can do that out of the box.